### PR TITLE
ISPN-13165 AsyncOfflineTest.testReset random failures (take 2)

### PR DESCRIPTION
**Pull request by pruivo**
Problems:
* State.accept() and onSendingCompleted() were running concurrently.
* onSendingCompleted() triggers an IRAC round but the State.stateStatus
   was not updated.
* The new round would not retry to send the key again making the test fail.

https://issues.redhat.com/browse/ISPN-13165

:crossed_fingers:

### CODE REVIEW

**Review by danberindei — COMMENTED** [review]
Sorry for the late review, forgot to press the green button

**Comment by pruivo** [comment]
updated

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 1]
```diff
@@ -517,6 +513,40 @@ boolean isExpiration() {
...
+            if (stateStatus != StateStatus.SENDING) {
+               // discarded or overwritten by another write operation
+               // do not send the cleanup command.
+               return;
+            }
+            stateStatus = StateStatus.COMPLETED;
+         }
+         onRoundCompleted(IracResponseCollector.Result.OK, false);
+         removeStateFromCluster(this);
+      }
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:531]: This isn't exactly what I had in mind... My point was that I hate it code that does similar things should live in methods with similar names to let the user know that they sit at the same abstraction level.

You already have methods `discard()` and `sendFail()`, which only update the `stateStatus` and log, so I was suggesting

```suggestion
      synchronized void sendSuccess() {
         // send succeed.
         if (log.isTraceEnabled()) {
            log.tracef("[IRAC] State.onSuccess for key %s (status=%s)", key, stateStatus);
         }
         if (stateStatus != StateStatus.SENDING) {
            // discarded or overwritten by another write operation
            // do not send the cleanup command.
            return;
         }
         stateStatus = StateStatus.COMPLETED;
      }
```

The two lines at the end of `onSuccess()` would remain in the `onResponse()` method.

**Comment by pruivo** [comment]
updated
* renamed `sendFail()` with `retrySend()`
* on `discard()` method removed `removeStateFromLocal(this)` from the synchronized block

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 2]
```diff
@@ -496,17 +492,19 @@ synchronized boolean canSend() {
          return false;
       }
 
-      synchronized void discard() {
-         if (log.isTraceEnabled()) {
-            log.tracef("[IRAC] State.onDiscard for key %s (status=%s)", key, stateStatus);
+      void discard() {
+         synchronized (this) {
+            if (log.isTraceEnabled()) {
+               log.tracef("[IRAC] State.onDiscard for key %s (status=%s)", key, stateStatus);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:498]:
```suggestion
               log.tracef("[IRAC] State.diiscard for key %s (status=%s)", key, stateStatus);
```

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 1 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:533]: Yeah, I missed that!
It still bothers me that `onRoundCompleted()` is called both from onResponse and from onResponse->onSuccess, but I can live with it ;)

**Comment by pruivo** [comment]
updated!

**Comment by danberindei** [comment]
Merged, thanks Pedro!

**Pull request merged**